tests/python/test_ad_global_data_access_rule_checker.py

```diff
@@ -134,3 +135,38 @@ def kernel_2_grad():
 
     with ti.ad.Tape(loss=loss, validation=True):
         kernel_2()
+
+
+@test_utils.test(require=ti.extension.assertion,
+                 exclude=[ti.cc],
+                 debug=True,
+                 validate_autodiff=True)
+def test_autodiff_mode_recovered():
+    N = 16
+    x = ti.field(dtype=ti.f32, shape=N, needs_grad=True)
+    loss = ti.field(dtype=ti.f32, shape=(), needs_grad=True)
+    b = ti.field(dtype=ti.f32, shape=(), needs_grad=True)
+
+    @ti.kernel
+    def kernel_1():
+        loss[None] = x[1] * b[None]
+
+    @ti.kernel
+    def kernel_2():
+        loss[None] = x[1] * b[None]
+
+    for i in range(N):
+        x[i] = i
+
+    b[None] = 10
+    loss.grad[None] = 1
+
+    func_calls = []
+    with ti.ad.Tape(loss=loss, validation=True) as t:
+        kernel_1()
+        kernel_2()
+        for f, _ in t.calls:
+            assert f.autodiff_mode == AutodiffMode.VALIDATION
+        func_calls = t.calls
+    for f, _ in func_calls:
```

Comment:
nit: can we have another test case that autodiff_mode before entering the context manager is not AutodiffMode.NONE to make sure we restore them correctly?